File: LM 2019/MedHomeward.java
// Author: Daniel Foster
// Difficulty: Medium - standard use of hashmaps.
// Needed: Hashmaps.

import java.util.*;

public class MedHomeward {
    public static void main(String [] args) {
        Scanner scan = new Scanner(System.in);

        // Test cases.
        int cases = scan.nextInt();
        while (cases-- > 0) {
            // Maps arrival city to departure city.
            Map<String, String> trips = new HashMap<>();

            // Number of boarding passes.
            int x = scan.nextInt();
            while (x-- > 0) {
                // Map new arrival to new departure.
                String dep = scan.next(), arr = scan.next();
                trips.put(arr, dep);
            }

            // Find final destination.
            String last = "";
            for (String s : trips.keySet()) {
                if (!trips.containsValue(s)) {
                    last = s;
                    break;
                }
            }

            // Loop back through the passes until original is reached.
            System.out.println(last);
            while (trips.containsKey(last)) {
                last = trips.get(last);
                System.out.println(last);
            }
        }
    }
}

File: LM 2019/MedHomeward.py
# Author: Daniel Foster
# Difficulty: Medium - standard use of dictionaries.
# Needed: Dictionaries.

# Test cases.
for _ in range(int(input())):
    # Maps arrival city to departure city.
    trips = {}

    # Read all boarding passes.
    for _ in range(int(input())):
        dep, arr = input().split()
        trips[arr] = dep
    
    # Find final destination.
    for s in trips:
        if s not in trips.values():
            last = s
            break
    
    # Loop back through passes until original is reached.
    print(last)
    while last in trips:
        last = trips[last]
        print(last)
